### PR TITLE
fix of issue 1479

### LINKED ISSUE #1479: Crash when pressing UP key in AutoCompleteBox hint
Take a look at the video: [AutoCompleteBox-PressUpInHint-Bug.zip](https://github.com/AvaloniaUI/Avalonia/files/1888894/AutoCompleteBox-PressUpInHint-Bug.zip)

### PR DESCRIPTION
**Pull request by achal7**
1)  changed UpdateLastState to call UpdateLastState(TState) so that it is the only one that has last update control
2) Clear function clean up the states, however currentNode by definition is depends upon one of the state value, hence made that also null,
since there is nothing now to link

Fixes #1479

### CODE REVIEW

File: src/Avalonia.Controls/TextBlock.cs [hunk 1]
```diff
@@ -364,14 +364,21 @@ protected virtual FormattedText CreateFormattedText(Size constraint)
...
         /// </summary>
         protected void InvalidateFormattedText()
+        {
+            ClearFormattedText();
+            InvalidateMeasure();
+        }
+
+        /// <summary>
+        /// Invalidates <see cref="FormattedText"/>.
+        /// </summary>
```
Comment (grokys) [inline src/Avalonia.Controls/TextBlock.cs]: I'm not so keen on having both `InvalidateFormattedText`  and `ClearFormattedText`. Could we just have `InvalidateFormattedText` and then explicitly call `InvalidateMeasure` when needed?

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public TextPresenter()
             Observable.Merge(
                 this.GetObservable(SelectionStartProperty),
                 this.GetObservable(SelectionEndProperty))
-                .Subscribe(_ => InvalidateFormattedText());
+                .Subscribe(_ => ClearFormattedText());
```
Comment (grokys) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs]: Actually, is this change related to fixing #1479? Or did you just see an opportunity to avoid an unnecessary layout pass?

**Review by grokys — COMMENTED** [review]
Looks good, just a couple of things.

File: src/Avalonia.Controls/TextBlock.cs (hunk 1 above)
Comment (achal7) [reply src/Avalonia.Controls/TextBlock.cs]: Yes, thats possible. I have kept those 2 function gets used in same call in order to not to break those who are already using that function. To implement your proposal we also need to update thoose places who are usnig InvalidateFormattedText to also call InvalidateMeasure. 
Then we dont need a function who do both

Comment (achal7) [reply src/Avalonia.Controls/TextBlock.cs]: Let me know if you would like to do that or not, if so I update those who are right now using InvalidateFormattedText

File: src/Avalonia.Controls/Presenters/TextPresenter.cs (hunk 2 above)
Comment (achal7) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: This one by fix of #1448.
By mistake this pull request combined 2 issue fixes, if you want i withdrow this PR and create 2 new PRs each targetting 2 different issues.

Regarding #1448 fix, coping comment here:

InvalidateFormattedText was doing mainly 2 activities, resets formatted text and resets measure.
Splited it and extracted reset of formatted text in to seprate function and using it over there.

TextPresenter usis obervable stream for SelectionStart/End property, over there inside subscriber instead of calling InvalidateFormatedText now we use only ClearFormatedText.

The reason is SelectionStart/End is not going to make any changes regarding measurement and hence reseting formatted text alone is sufficient enough

Comment (grokys) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: Ah yes, I remember you saying. Hmm yes it would be better if these were 2 separate PRs really, it's hard to discuss two changes when they're mixed like this. You don't need to close this one, just create a new branch for the other PR, do a `git reset` to the correct commit, followed by a force push and then create a PR from the other branch, if that makes sense.

File: src/Avalonia.Controls/TextBlock.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Controls/TextBlock.cs]: I'm fine with changing the behavior of `InvalidateFormattedText` at this point - we're still in beta, we can do stuff like that ;) So yeah go ahead and make the change.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs (hunk 2 above)
Comment (achal7) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: No luck to split the commits. I have created new branch in my fork fixes/1479 and made the changes and created new PR.
Created another branch fixes/1448 and now making the changes regarding ResetFormatedText  and once it is done, will create seperate PR.

Comment (achal7) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: No luck to split the commits. I have created new branch in my fork fixes/1479 and made the changes and created new PR.
Created another branch fixes/1448 and now making the changes regarding ResetFormatedText  and once it is done, will create separate PR.

**Comment by achal7** [comment]
This PR now only for #1479 
For #1448 created new PR #1512

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Thanks @achal7 !
